PR 신청합니다

InJun2 opened the pull request:
- Todo-List CRUD 기능 추가
- 미흡한 부분이나 추가해야하는 부분이 존재하면 빠르게 조치하겠습니다..

InJun2 committed:
- JPA CRUD 진행
- todo 생성/전체조회/페이지조회/단일삭제/단일변경 기능
- 그외 설정 존재하지 않음

InJun2 committed:
- TODO CRUD 기능

InJun2 committed:
- README 간단한 기능 정리

InJun2 committed:
Merge remote-tracking branch 'origin/main'

DongGeon0908 commented on src/main/java/kr/co/selab/selabtodo/controller/TodoController.java:
```diff
@@ -0,0 +1,43 @@
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("api/todo")
```
/api/todo/v1

DongGeon0908 replied:
version을 추가하는게 좋아요 언제 api가 변경될지 모르니까
그리고 문서화를 위해서 swagger를 추가하는 건 어떨까요?

DongGeon0908 commented on src/main/java/kr/co/selab/selabtodo/entity/TodoEntity.java:
```diff
@@ -0,0 +1,53 @@
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Entity
+@Table(name="TODOLIST")
```
sql 표기법 확인 부탁드립니다

DongGeon0908 commented on src/main/java/kr/co/selab/selabtodo/service/TodoService.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Transactional(readOnly = true)
+    public TodoResponseDto searchTodo(long id){
+        try {
+            return repository.findById(id).orElseThrow(RelationNotFoundException::new).toTodoResponseDto();
+        } catch (RelationNotFoundException e) {
+            throw new RuntimeException(e);
+        }
```
try-catch가 아니라 Exception 핸들링하는게 맞아요
몇가지 키워드를 알려드릴게요 찾아보시길
RestControllerAdvice
ExceptionHandler

DongGeon0908 reviewed (changes requested):
고생하셨습니다 몇가지 수정작업을 진행하면 될 것 같아요

1. API 문서화 진행
2. Exception 핸들링 추가

InJun2 committed:
- Exception 추가
- DB 테이블명 변경
- swagger 적용

DongGeon0908 commented on src/main/java/kr/co/selab/selabtodo/controller/TodoController.java:
```diff
@@ -0,0 +1,51 @@
+    @Operation(summary = "Create Todo-List", description = "title과 content 요청 값으로 Todo List 생성")
+    @PostMapping
+    public ResponseEntity<TodoResponseDto> createTodo(@RequestBody CreateTodoRequestDto todo){
+        return ResponseEntity.ok(service.createTodo(todo));
```
Http Status Code에서 생성할 때는 어떤 상태값을 가져갈가요?

DongGeon0908 commented on src/main/java/kr/co/selab/selabtodo/controller/TodoController.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Operation(summary = "Update Todo-List", description = "id를 통한 조회 및 Todo-List 업데이트")
+    @PatchMapping("/{id}")
+    public ResponseEntity<TodoResponseDto> updateTodo(@PathVariable long id, @RequestBody UpdateTodoRequestDto todo) throws RelationNotFoundException {
```
controller까지 throw가 넘어올 필요가 있을까요?

DongGeon0908 commented on src/main/java/kr/co/selab/selabtodo/entity/TodoEntity.java:
```diff
@@ -0,0 +1,53 @@
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Entity
+@Table(name="TODO_LIST")
```
todo_list

DongGeon0908 commented on src/main/java/kr/co/selab/selabtodo/service/TodoService.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    @Transactional
+    public TodoResponseDto updateTodo(long id, UpdateTodoRequestDto todo) throws RelationNotFoundException {
```
Exception을 왜 throw할까요?

DongGeon0908 commented on src/main/java/kr/co/selab/selabtodo/service/TodoService.java:
```diff
@@ -0,0 +1,52 @@
+    private final TodoRepository repository;
+
+    @Transactional(readOnly = true)
+    public TodoResponseDto searchTodo(long id) throws RelationNotFoundException {
```
Exception을 왜 throw할까요?

DongGeon0908 reviewed (changes requested):
1. http status code에 맞게 구성하기
2. Exception 처리 명확하게
3. 읽기 쉬운 코드를 위해 단락을 구분짓기 (한줄로 표현하는 것보다는 줄을 띄어서 표현)

InJun2 committed:
- Http Status code 수정
- Exception 수정

DongGeon0908 commented on src/main/java/kr/co/selab/selabtodo/controller/TodoController.java:
```diff
@@ -21,25 +23,29 @@
+        return ResponseEntity.status(HttpStatus.CREATED)
+                .body(service.updateTodo(id,todo));
     }
 
```
업데이트도 status code가 201인가요?

DongGeon0908 commented on src/main/java/kr/co/selab/selabtodo/controller/TodoController.java:
```diff
@@ -0,0 +1,57 @@
+        return ResponseEntity.status(HttpStatus.CREATED)
+                .body(service.createTodo(todo));
+    }
+
+    @Operation(summary = "Update Todo-List", description = "id를 통한 조회 및 Todo-List 업데이트")
+    @PatchMapping("/{id}")
+    public ResponseEntity<TodoResponseDto> updateTodo(@PathVariable @Positive long id,
+                                                      @RequestBody UpdateTodoRequestDto todo) {
+        return ResponseEntity.status(HttpStatus.CREATED)
+                .body(service.updateTodo(id,todo));
+    }
```
수정도 상태코드가 201인가요?

DongGeon0908 commented on src/main/java/kr/co/selab/selabtodo/controller/TodoController.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    @Operation(summary = "Search All Todo-List", description = "모든 Todo-List 조회")
+    @GetMapping("/all")
```
"/all" 필요없을 듯 합니다

DongGeon0908 commented on src/main/java/kr/co/selab/selabtodo/exception/ExceptionHandlers.java:
```diff
@@ -0,0 +1,31 @@
+@RestControllerAdvice
+public class ExceptionHandlers {
+
+    @ExceptionHandler(NotExistTodoException.class)
+    public ResponseEntity handleRelationNotFoundException(NotExistTodoException e){
+        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(
+                ErrorResponse.builder()
+                        .code(404)
+                        .message(e.getMessage())
+                        .build()
+        );
+    }
+
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity handleMethodArgumentNotValidException(MethodArgumentNotValidException e){
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(
+                ErrorResponse.builder()
+                        .code(400)
+                        .message(e.getMessage())
+                        .build()
+        );
+    }
```
error에 대한 로깅을 추가해주는 건 어떨까요?

DongGeon0908 commented on src/main/java/kr/co/selab/selabtodo/service/TodoService.java:
```diff
@@ -0,0 +1,58 @@
+
+    @Transactional(readOnly = true)
+    public InquiryTodoPageResponseDto searchPageTodos(Pageable page) {
+        return new InquiryTodoPageResponseDto(page.getPageNumber(),
+                repository.findAll(page).stream()
+                .map(TodoEntity::toTodoResponseDto)
+                .collect(Collectors.toList()));
```
한번 나눠주는 건 어떨까요?
```suggestion
        var response = page.getPageNumber(),
                repository.findAll(page).stream()
                .map(TodoEntity::toTodoResponseDto)
                .collect(Collectors.toList())
        
        return new InquiryTodoPageResponseDto(response);
```

DongGeon0908 replied:
좋은 코드는 상대방이 이해하기 편한 코드라고 생각해요

DongGeon0908 reviewed (changes requested):
- CORS 문제는 어떻게 해결하실건가요?
- 더티체킹을 통해 업데이트를 진행하는 건 어떨까요?

InJun2 committed:
- 업데이트 Http Status code 수정
- update 더티체킹으로 수정, 쓸모없는 빌더 메서드 삭제
- 스트림 사용 메서드 var 변수를 사용하여 분리
- WebConfig를 이용하여 전역적으로 Cors 문제 해결

DongGeon0908 commented on src/main/java/kr/co/selab/selabtodo/controller/TodoController.java:
```diff
@@ -0,0 +1,56 @@
+import javax.validation.constraints.PositiveOrZero;
+
+@RestController
+@RequestMapping("api/todo/v1")
```
api/v1/todo가 더 좋을 것 같네요

DongGeon0908 commented on src/main/java/kr/co/selab/selabtodo/controller/TodoController.java:
```diff
@@ -0,0 +1,56 @@
+    @DeleteMapping("/{id}")
+    public ResponseEntity<Long> deleteTodoById(@PathVariable @Positive long id){
+        service.deleteTodoById(id);
+        return ResponseEntity.ok(id);
```
noContentBody

DongGeon0908 commented on src/main/java/kr/co/selab/selabtodo/dto/TodoResponseDto.java:
```diff
@@ -0,0 +1,22 @@
+@Getter
+@NoArgsConstructor
+public class TodoResponseDto {
+    Long id;
+    String title;
+    String content;
+    boolean checked;
```
접근제한자는 어디에

DongGeon0908 reviewed (commented):
이정도면 된 것 같아요 
그런데 몇가지 내용 확인 해주세요

1. 일관성 있게 코드를 작성해주세요. 어디서는 var, 어드서든 원시형... 통일 필수
2. 읽기 쉬운 코드가 좋은 코드에요 지금은 읽기 어려워요 정렬 필수
3. 기존 리뷰에 대해 한번더 고민해보면 좋을 듯 합니다
